packages/frontend/src/context/StoreContext.tsx

```diff
@@ -202,13 +213,40 @@ export const StoreContextProvider = (
             setErc20Addr,
             setPublishedTagId,
             setFinalizedCarts,
+            setPubKeys,
+            walletAddress,
           );
         }
       }
     } catch (err) {
       console.error("error receiving events", err);
     }
   };
+  const verify = async (
```

Comment:
calling verifySignedTypeData for every cartId (that have not been invalidated/failed) with every key card address associated with current wallet.

Reply:
> every key card address associated with current wallet

Can you expand on this? The event might be signed by some other user, too.

Basically you get the keyCard/public key that signed the event via `ecrecover()` (in python it's called `recover_message`) and then you just need to check if the key is in the set of KeyCards + the assigned relays from the contract. 

```py
    def _valid_event_signing_addresses(self):
        all = []
        # retreive all relays nfts
        if self.storeReg.functions.getRelayCount(self.store_token_id).call() > 0:
            all_relay_token_ids = self.storeReg.functions.getAllRelays(self.store_token_id).call()
            for token_id in all_relay_token_ids:
                # retreive the owner => it's address
                relay_address = self.relayReg.functions.ownerOf(token_id).call()
                all.append(relay_address.lower())
        # turn key cards into addresses
        key_card_addresses = [public_key_to_address(pk).lower() for pk in list(self.all_key_cards.keys())]
        all.extend(key_card_addresses)
        return all

    def _verify_event(self, evt: schema_pb2.Event):
        if len(evt.signature) != 65:
            raise Exception("Invalid signature length: {}".format(len(evt.signature)))
        encoded_data = self.event_hasher.hash_event(evt)
        pub_key = self.w3.eth.account.recover_message(encoded_data, signature=evt.signature)
        print(f"event:{get_id_from_event(evt)} recovered pub_key: {pub_key}")
        their_addr = Web3.to_checksum_address(pub_key).lower()
        valid_addrs = self._valid_event_signing_addresses()
        if their_addr not in valid_addrs:
            print(f"valid addresses: {valid_addrs}")
            raise Exception("Event signed by unknown address: {}".format(their_addr))
```

`self.all_key_cards` should be very similar to the reducer you built.

Main thing is, you shouldn't have to iterate over all the possible signing addresses again and again.